x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/nulls/Coalesce.java

```diff
@@ -0,0 +1,158 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.scalar.nulls;
+
+import org.elasticsearch.compute.data.Block;
+import org.elasticsearch.compute.data.ElementType;
+import org.elasticsearch.compute.data.Page;
+import org.elasticsearch.compute.operator.EvalOperator;
+import org.elasticsearch.xpack.esql.planner.LocalExecutionPlanner;
+import org.elasticsearch.xpack.esql.planner.Mappable;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Nullability;
+import org.elasticsearch.xpack.ql.expression.TypeResolutions;
+import org.elasticsearch.xpack.ql.expression.function.scalar.ScalarFunction;
+import org.elasticsearch.xpack.ql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+
+import static org.elasticsearch.xpack.ql.type.DataTypes.NULL;
+
+/**
+ * Function returning the first non-null value.
+ */
+public class Coalesce extends ScalarFunction implements Mappable {
+    private DataType dataType;
+
+    public Coalesce(Source source, List<Expression> expressions) {
+        super(source, expressions);
+    }
+
+    @Override
+    public DataType dataType() {
+        if (dataType == null) {
+            resolveType();
+        }
+        return dataType;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        for (int position = 0; position < children().size(); position++) {
+            if (dataType == null || dataType == NULL) {
+                dataType = children().get(position).dataType();
+                continue;
+            }
+            TypeResolution resolution = TypeResolutions.isType(
+                children().get(position),
+                t -> t == dataType,
+                sourceText(),
+                TypeResolutions.ParamOrdinal.fromIndex(position),
+                dataType.typeName()
+            );
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    @Override
+    public Nullability nullable() {
+        return children().get(children().size() - 1).nullable();
+    }
+
+    @Override
+    public ScriptTemplate asScript() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new Coalesce(source(), newChildren);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Coalesce::new, children());
+    }
+
+    @Override
+    public boolean foldable() {
```

Comment:
In ES SQL there is a Coalesce function which is interesting because it's also coming together with an Optimizer rule and the `foldable` method is much simpler (thanks to the aforementioned rule).

Otherwise, there is `Expressions.foldable(children());` that can be used here.

The way I see it:
- either you merge this as is and create a GH issue to explore/improve NULL propagation for Coalesce (plus that optimizer rule)
- have a look at Coalesce in ES SQL and SQL's `Optimizer.PropagateNullable` and have an adapted version in ESQL

Reply:
I'll open #98612 and replace with `Expressions.foldable(children())` and we'll grab an optimizer rule later.